File: 2020/day11/sol.cpp
#include <fstream>
#include <iostream>
#include <vector>
#include <algorithm>

bool valid_coordinate(const int x, const int y, const int width, const int height) {
    return 0 <= x && x < width && 0 <= y && y < height;
}

size_t get_nearby_seats(const std::vector<std::string>& seats, const int x, const int y) {
    size_t nearby_seats = 0;
    int line_length = seats[0].length();
    int number_of_lines = seats.size();

    for (int rx = -1; rx <= 1; ++rx) {
        for (int ry = -1; ry <= 1; ++ry) {
            if (rx == 0 && ry == 0) {
                continue;
            }
            int current_x = x + rx;
            int current_y = y + ry;
            if (valid_coordinate(current_x, current_y, line_length, number_of_lines)) {
                if (seats[current_y][current_x] == '#') {
                    nearby_seats++;
                }
            }
        }
    }

    return nearby_seats;
}

size_t get_visible_seats(const std::vector<std::string>& seats, const int x, const int y) {
    size_t visible_seats = 0;
    int line_length = seats[0].length();
    int number_of_lines = seats.size();

    for (int rx = -1; rx <= 1; ++rx) {
        for (int ry = -1; ry <= 1; ++ry) {
            int current_x = x + rx;
            int current_y = y + ry;
            if (rx == 0 && ry == 0) {
                continue;
            }
            while (valid_coordinate(current_x, current_y, line_length, number_of_lines)) {
                if (seats[current_y][current_x] == '#') {
                    visible_seats++;
                    break;
                } else if (seats[current_y][current_x] == 'L') {
                    break;
                } else {
                    current_x += rx;
                    current_y += ry;
                }
            }
        }
    }

    return visible_seats;
}

template <typename T>
bool vec_equal(const std::vector<T>& a, const std::vector<T>& b) {
    if (a.size() != b.size()) {
        return false;
    }
    return std::equal(a.begin(), a.end(), b.begin());
}

size_t occupied_seats(std::vector<std::string> seats,
                      size_t (*f)(
                          const std::vector<std::string>&,
                          const int, const int
                          ),
                      const size_t tolerancy_limit) {
    size_t occupied_seats = 0;
    std::vector<std::string> seats_swap;

    seats_swap = seats;

    while (true) {
        for (size_t y = 0; y < seats.size(); ++y) {
            for (size_t x = 0; x < seats[0].length(); ++x) {
                if (seats_swap[y][x] == 'L' && f(seats_swap, x, y) == 0) {
                    seats[y][x] = '#';
                } else if (seats_swap[y][x] == '#' && f(seats_swap, x, y) >= tolerancy_limit) {
                    seats[y][x] = 'L';
                }
            }
        }

        if (vec_equal(seats, seats_swap)) {
            break;
        }

        seats_swap = seats;
    }

    for (size_t y = 0; y < seats.size(); ++y) {
        for (size_t x = 0; x < seats[0].length(); ++x) {
            if (seats[y][x] == '#') {
                occupied_seats++;
            }
        }
    }

    return occupied_seats;
}

int main() {
    std::ifstream input("input");
    if (!input.is_open()) {
        std::cerr << "Can't read puzzle input file.\n";
        return -1;
    }

    std::vector<std::string> seats;

    std::string line;
    while (std::getline(input, line)) {
        seats.push_back(line);
    }

    std::cout << "part 1: " << occupied_seats(seats, get_nearby_seats, 4) << '\n';
    std::cout << "part 2: " << occupied_seats(seats, get_visible_seats, 5) << '\n';

    return 0;
}

File: 2020/day11/sol.rs
use std::convert::TryInto;
use std::fs;

fn valid_coordinates(x: i32, y: i32, width: i32, height: i32) -> bool {
    0 <= x && x < width && 0 <= y && y < height
}

fn get_nearby_seats(seats: &Vec<Vec<u8>>, x: i32, y: i32) -> usize {
    let mut nearby_seats = 0;

    for rx in -1..=1 {
        for ry in -1..=1 {
            if rx == 0 && ry == 0 {
                continue;
            }
            let current_x = x + rx;
            let current_y = y + ry;

            if let (Ok(nx), Ok(ny), true) = (
                TryInto::<usize>::try_into(current_x),
                TryInto::<usize>::try_into(current_y),
                valid_coordinates(
                    current_x,
                    current_y,
                    seats[0].len() as i32,
                    seats.len() as i32,
                ),
            ) {
                if seats[ny][nx] == b'#' {
                    nearby_seats += 1;
                }
            }
        }
    }

    nearby_seats
}

fn get_visible_seats(seats: &Vec<Vec<u8>>, x: i32, y: i32) -> usize {
    let mut visible_seats = 0;

    for rx in -1..=1 {
        for ry in -1..=1 {
            if rx == 0 && ry == 0 {
                continue;
            }
            let mut current_x = x + rx;
            let mut current_y = y + ry;

            while let (Ok(nx), Ok(ny), true) = (
                TryInto::<usize>::try_into(current_x),
                TryInto::<usize>::try_into(current_y),
                valid_coordinates(
                    current_x,
                    current_y,
                    seats[0].len() as i32,
                    seats.len() as i32,
                ),
            ) {
                if seats[ny][nx] == b'#' {
                    visible_seats += 1;
                    break;
                } else if seats[ny][nx] == b'L' {
                    break;
                } else {
                    current_x += rx;
                    current_y += ry;
                }
            }
        }
    }

    visible_seats
}

fn occupied_seats(
    mut seats: Vec<Vec<u8>>,
    f: fn(&Vec<Vec<u8>>, i32, i32) -> usize,
    tolerancy_limit: usize,
) -> usize {
    let mut seats_swap = seats.clone();

    loop {
        for y in 0..seats.len() {
            for x in 0..seats[0].len() {
                let adjacent = f(&seats_swap, x as i32, y as i32);
                if seats_swap[y][x] == b'L' && adjacent == 0 {
                    seats[y][x] = b'#';
                } else if seats_swap[y][x] == b'#' && adjacent >= tolerancy_limit {
                    seats[y][x] = b'L';
                }
            }
        }

        if seats == seats_swap {
            break;
        }

        seats_swap = seats.clone();
    }

    seats
        .iter()
        .map(|s| s.iter().filter(|&c| c == &b'#').count())
        .sum()
}

fn main() {
    let input = fs::read_to_string("input").expect("Can't read puzzle input file.");

    let seats = input
        .lines()
        .map(|s| s.as_bytes().to_vec())
        .collect::<Vec<_>>();

    println!(
        "part 1: {}",
        occupied_seats(seats.clone(), get_nearby_seats, 4)
    );
    println!("part 2: {}", occupied_seats(seats, get_visible_seats, 5));
}
